contrib: allow multi-sig binary verification v2

theuni opened the pull request:
Following up on #23020 from jamesob with achow101's additional features on top. 

Both mentioned that they will be away for the next few weeks, so this is intended to keep review going.

All credit to the jamesob and achow101. See #23020 for the original description and [here](https://github.com/bitcoin/bitcoin/pull/23020#issuecomment-1480603300) for the added features.

I squashed the last commit from https://github.com/achow101/bitcoin/tree/pr23020-direct-bins-gpg-parse into the first commit here.

Fetching and local verification seem to work as intended for me.

jamesob committed:
contrib: verifybinaries: allow multisig verification

This commit adds the functionality necessary to transition from
doing binary verification on the basis of a single signature to
requiring a minimum threshold of trusted signatures.

A signature can appear as "good" from GPG output, but it may not come
from an identity the user trusts. We call these "good, untrusted"
signatures.

We report bad signatures but do not necessarily fail in their presence,
since a bad signature might coexist with enough good, trusted signatures
to fulfill our criteria.

If "--import-keys" is enabled, we will prompt the user to
optionally try to retrieve unknown keys. Marking them as trusted locally
is a WIP, but keys which are retrieved successfully and appear on the
builder-keys list will immediately count as being useful towards
fulfilling the threshold.

Logging is improved and an option to output JSON that summarizes the
whole sum signature and binary verification processes has been added.

Co-authored-by: Russ Yanofsky <russ@yanofsky.org>
Co-authored-by: willcl-ark <will8clark@gmail.com>

achow101 committed:
contrib: Refactor verifbinaries to support subcommands

Prepares for the option to provide local binaries, sha256sums, and
signatures directly.

achow101 committed:
contrib: Specify to GPG the SHA256SUMS file that is detached signed

achow101 committed:
contrib: Add verifybinaries command for specifying files to verify

In addition to verifying the published releases with the `pub` command,
the verifybinaries script is updated to take a `bin` command where the
user specifies the local files, sums, and sigs to verify.

achow101 committed:
contrib: Use machine parseable GPG output in verifybinaries

GPG has an option to provide machine parseable output. Use that instead
of trying to parse the human readable output.

hebasto commented on contrib/verifybinaries/README.md:
```diff
@@ -1,30 +1,77 @@
+You can obtain these keys by
+  - through a browser using a key server (e.g. keyserver.ubuntu.com),
+  - manually using the `gpg --keyserver <url> --recv-keys <key>` command, or
+  - you can run the packaged `verifybinaries.py ... --import-keys` script to 
```
2e187ffafc9cb0490eedd079ce1cfffafecd9dba: trailing whitespace
```suggestion
  - you can run the packaged `verifybinaries.py ... --import-keys` script to
```

TheCharlatan commented:
> Fetching and local verification seem to work as intended for me.

Can you post your commands? Nothing is currently working as advertised in the README for me ^^.

willcl-ark commented:
@TheCharlatan I see it working using e.g. the `pub` subcommand. You need to choose `pub` or `bin` to verify either the publicly-hosted binaries, or `bin` for a local binary. But you are correct the docs are now outdated...

```fish
₿ ./contrib/verifybinaries/verify.py pub bitcoin-core-23.0-x86_64
[INFO] got file https://bitcoincore.org/bin/bitcoin-core-23.0/SHA256SUMS.asc as SHA256SUMS.asc
[WARNING] https://bitcoin.org failed to provide file (https://bitcoin.org/bin/bitcoin-core-23.0/SHA256SUMS.asc). Continuing based solely upon https://bitcoincore.org.
[INFO] got file https://bitcoincore.org/bin/bitcoin-core-23.0/SHA256SUMS as SHA256SUMS
[WARNING] https://bitcoin.org failed to provide file (https://bitcoin.org/bin/bitcoin-core-23.0/SHA256SUMS). Continuing based solely upon https://bitcoincore.org.
[INFO] got 17 good signatures
[INFO] GOOD SIGNATURE (untrusted): SigData('17565732E08E5E41', 'Andrew Chow <andrew@achow101.com>', trusted=False, status='')
[INFO] GOOD SIGNATURE (untrusted): SigData('D7CC770B81FD22A8', 'Ben Carman <benthecarman@live.com>', trusted=False, status='expired')
[INFO] GOOD SIGNATURE (untrusted): SigData('E13FC145CD3F4304', 'Antoine Poinsot <darosior@protonmail.com>', trusted=False, status='')
[INFO] GOOD SIGNATURE (untrusted): SigData('ED357015286A333D', 'Duncan Dean <duncangleeddean@gmail.com>', trusted=False, status='')
[INFO] GOOD SIGNATURE (untrusted): SigData('3152347D07DA627C', 'Stephan Oeste (it) <it@oeste.de>', trusted=False, status='')
[INFO] GOOD SIGNATURE (untrusted): SigData('2EEB9F5CC09526C1', 'Michael Ford (bitcoin-otc) <fanquake@gmail.com>', trusted=False, status='')
[INFO] GOOD SIGNATURE (untrusted): SigData('8E4256593F177720', 'Oliver Gugger <gugger@gmail.com>', trusted=False, status='')
[INFO] GOOD SIGNATURE (untrusted): SigData('410108112E7EA81F', 'Hennadii Stepanov (GitHub key) <32963518+hebasto@users.noreply.github.com>', trusted=False, status='')
[INFO] GOOD SIGNATURE (untrusted): SigData('D11BD4F33F1DB499', 'jackielove4u <jackielove4u@hotmail.com>', trusted=False, status='')
[INFO] GOOD SIGNATURE (untrusted): SigData('DD78544CF91B1514', 'Jonas Ott (jonas-ott) <git@jonas-ott.de>', trusted=False, status='')
[INFO] GOOD SIGNATURE (untrusted): SigData('F62711DBDCA8AE56', 'Dimitri <kvaciral@protonmail.com>', trusted=False, status='')
[INFO] GOOD SIGNATURE (untrusted): SigData('1E4AED62986CD25D', 'Wladimir J. van der Laan <laanwj@protonmail.com>', trusted=False, status='')
[INFO] GOOD SIGNATURE (untrusted): SigData('BD02942421F4889F', 'Luke Dashjr <luke@dashjr.org>', trusted=False, status='revoked')
[INFO] GOOD SIGNATURE (untrusted): SigData('0A41BDC3F4FAFF1C', 'Aaron Clauson (sipsorcery) <aaron@sipsorcery.com>', trusted=False, status='')
[INFO] GOOD SIGNATURE (untrusted): SigData('7E296D555E7F63A7', 'Sjors Provoost <sjors@purpledunes.com>', trusted=False, status='')
[INFO] GOOD SIGNATURE (untrusted): SigData('A7BEB2621678D37D', 'vertion <vertion@protonmail.com>', trusted=False, status='expired')
[INFO] GOOD SIGNATURE (untrusted): SigData('099BAD163C70FBFA', 'Will Clark <will8clark@gmail.com>', trusted=False, status='')
[WARNING] key D7CC770B81FD22A8 for Ben Carman <benthecarman@live.com> is expired
[WARNING] key A7BEB2621678D37D for vertion <vertion@protonmail.com> is expired
[INFO] removing *-unsigned binaries (bitcoin-23.0-arm64-apple-darwin-unsigned.dmg, bitcoin-23.0-arm64-apple-darwin-unsigned.tar.gz, bitcoin-23.0-x86_64-apple-darwin-unsigned.dmg, bitcoin-23.0-x86_64-apple-darwin-unsigned.tar.gz, bitcoin-23.0-win64-setup-unsigned.exe, bitcoin-23.0-win64-unsigned.tar.gz) from verification since https://bitcoincore.org does not host *-unsigned binaries
[INFO] removing *-debug binaries (bitcoin-23.0-aarch64-linux-gnu-debug.tar.gz, bitcoin-23.0-arm-linux-gnueabihf-debug.tar.gz, bitcoin-23.0-powerpc64-linux-gnu-debug.tar.gz, bitcoin-23.0-powerpc64le-linux-gnu-debug.tar.gz, bitcoin-23.0-riscv64-linux-gnu-debug.tar.gz, bitcoin-23.0-x86_64-linux-gnu-debug.tar.gz, bitcoin-23.0-win64-debug.zip) from verification since https://bitcoincore.org does not host *-debug binaries
[INFO] removing *-codesignatures binaries (bitcoin-23.0-codesignatures-e36a046909ad.tar.gz) from verification since https://bitcoincore.org does not host *-codesignatures binaries
[INFO] downloading bitcoin-23.0-aarch64-linux-gnu.tar.gz
 [INFO] downloading bitcoin-23.0-arm-linux-gnueabihf.tar.gz
[INFO] downloading bitcoin-23.0-arm64-apple-darwin.dmg
[INFO] downloading bitcoin-23.0-arm64-apple-darwin.tar.gz
[INFO] downloading bitcoin-23.0.tar.gz
[INFO] downloading bitcoin-23.0-powerpc64-linux-gnu.tar.gz
[INFO] downloading bitcoin-23.0-powerpc64le-linux-gnu.tar.gz
[INFO] downloading bitcoin-23.0-riscv64-linux-gnu.tar.gz
[INFO] downloading bitcoin-23.0-x86_64-apple-darwin.dmg
[INFO] downloading bitcoin-23.0-x86_64-apple-darwin.tar.gz
[INFO] downloading bitcoin-23.0-x86_64-linux-gnu.tar.gz
[INFO] downloading bitcoin-23.0-win64-setup.exe
[INFO] downloading bitcoin-23.0-win64.zip
[INFO] did not clean up /tmp/bitcoin_verify_binaries.bitcoin-core-23.0-x86_64
VERIFIED: bitcoin-23.0-aarch64-linux-gnu.tar.gz
VERIFIED: bitcoin-23.0-arm-linux-gnueabihf.tar.gz
VERIFIED: bitcoin-23.0-arm64-apple-darwin.dmg
VERIFIED: bitcoin-23.0-arm64-apple-darwin.tar.gz
VERIFIED: bitcoin-23.0.tar.gz
VERIFIED: bitcoin-23.0-powerpc64-linux-gnu.tar.gz
VERIFIED: bitcoin-23.0-powerpc64le-linux-gnu.tar.gz
VERIFIED: bitcoin-23.0-riscv64-linux-gnu.tar.gz
VERIFIED: bitcoin-23.0-x86_64-apple-darwin.dmg
VERIFIED: bitcoin-23.0-x86_64-apple-darwin.tar.gz
VERIFIED: bitcoin-23.0-x86_64-linux-gnu.tar.gz
VERIFIED: bitcoin-23.0-win64-setup.exe
VERIFIED: bitcoin-23.0-win64.zip
```

TheCharlatan commented:
> @TheCharlatan I see it working using e.g. the pub subcommand. You need to choose pub or bin to verify either the publicly-hosted binaries, or bin for a local binary. But you are correct the docs are now outdated...

Mmh, shouldn't the sample command you posted only have verified x86_64 binaries?

theuni commented on contrib/verifybinaries/README.md:
````diff
@@ -1,30 +1,77 @@
+```sh
+./contrib/verifybinaries/verify.py 22.0
+./contrib/verifybinaries/verify.py 22.0-rc2
+./contrib/verifybinaries/verify.py bitcoin-core-23.0
````
This isn't clear to me either.

fanquake replied:
[I mentioned in the original PR, that I think this can just be dropped](https://github.com/bitcoin/bitcoin/pull/23020#discussion_r1117049036). We know we are downloading bitcoin-core, so I'm not sure why we need to support anything other than just passing in a version number.

theuni commented:
> > @TheCharlatan I see it working using e.g. the pub subcommand. You need to choose pub or bin to verify either the publicly-hosted binaries, or bin for a local binary. But you are correct the docs are now outdated...
> 
> Mmh, shouldn't the sample command you posted only have verified x86_64 binaries?

Mmm, yeah, this is awkward. @achow101's original changes forced the user to specify which binaries should be checked. I requested using the contents of the SHA256SUMS file instead.

The problem with teaching the `bin` command the version syntax from `pub` is that that syntax is pretty remote-url centric. So I'm not sure it makes sense to allow the user to use the same type of filter for `bin`.

Should we just go back to requiring the user to specify the bins? Downloading to a clean dir and using a wildcard should still work for tooling I think. @TheBlueMatt ?

fanquake commented on contrib/verifybinaries/README.md:
````diff
@@ -1,30 +1,77 @@
 
 ```sh
-./verify.py bitcoin-core-0.13.0 delete
+./contrib/verifybinaries/verify.py bitcoin-core-22.0 delete
````
From https://github.com/bitcoin/bitcoin/pull/23020#discussion_r1117050954
> Should we just add an explicit `--delete` argument to the script, rather than "append a random anything", as was proposed in https://github.com/bitcoin/bitcoin/pull/26985.

theuni replied:
Looks like this has been implemented as `pub --cleanup`. A quick test shows it working as intended. Will update the doc.

glozow commented on contrib/verifybinaries/README.md:
```diff
@@ -1,30 +1,77 @@
+./contrib/verifybinaries/verify.py 22.0
+./contrib/verifybinaries/verify.py 22.0-rc2
+./contrib/verifybinaries/verify.py bitcoin-core-23.0
+./contrib/verifybinaries/verify.py bitcoin-core-23.0-rc1
```
nit: it would be nice to have a `bin` example as well
```suggestion
./contrib/verifybinaries/verify.py pub bitcoin-core-23.0-rc1
./contrib/verifybinaries/verify.py bin ~/Downloads/SHA256SUMS ~/Downloads/bitcoin-24.0.1-x86_64-linux-gnu.tar.gz
```

glozow commented on contrib/verifybinaries/verify.py:
```diff
@@ -47,136 +172,441 @@ def parse_version_string(version_str):
+    if gpg_retval not in gpg_allowed_codes:
+        if gpg_retval == 1:
+            log.critical(f"Bad signature (code: {gpg_retval}).")
+        if gpg_retval == 2:
+            log.critical(
+                "gpg error. Do you have the Bitcoin Core binary release "
+                "signing key installed?")
```
nit: This seems like it can be deleted, since 2 is an allowed return code and missing "the Bitcoin Core binary release signing key" doesn't apply

```suggestion
```

theuni committed:
verifybinaries: move all current examples to the pub subcommand

theuni committed:
verifybinaries: remove awkward bitcoin-core prefix handling

theuni committed:
verifybinaries: Don't delete shasums file

It may be useful for local validation.

theuni committed:
verifybinaries: README cleanups

- Use correct name for verify.py
- Add usage examples for verifybinaries bin
- Document proper use of new cleanup option
- Fixup broken example

theuni commented on contrib/verifybinaries/README.md:
```diff
@@ -1,30 +1,77 @@
+You can obtain these keys by
+  - through a browser using a key server (e.g. keyserver.ubuntu.com),
+  - manually using the `gpg --keyserver <url> --recv-keys <key>` command, or
+  - you can run the packaged `verifybinaries.py ... --import-keys` script to
```
Not sure what you mean, I think this just intends to say `verify.py` ?

TheCharlatan replied:
I was inferring that there may be some undocumented step copying and renaming this file into the distributed tar archives for verification purposes. If this is not the case, then I think this indeed should just say `verify.py`.

theuni committed:
verifybinaries: remove unreachable code

theuni commented:
I believe I've addressed the comments here. I've chosen to add fresh comments on top rather than squashing to avoid mixing up my changes on top of @jamesob's/@achow101's.

The os/arch filters like `verify.py 22.0-osx` no longer work for me and I'm not quite sure why. Otherwise remote/local verification seem to work ok.

fanquake commented on contrib/verifybinaries/verify.py:
```diff
@@ -47,137 +115,598 @@ def parse_version_string(version_str):
+    )
+    parser.add_argument(
+        '--keyserver', action='store', nargs='?',
+        default=os.environ.get('BINVERIFY_KEYSERVER', 'hkp://keyserver.ubuntu.com'),
```
note that the keyserver we currently "recommend" is `hkps://keys.openpgp.org`. See #22688, #23466.

fanquake commented on contrib/verifybinaries/README.md:
````diff
@@ -1,30 +1,89 @@
+specified keys, while requiring a threshold of at least 10 trusted signatures:
+```sh
+./contrib/verifybinaries/verify.py pub 22.0-x86 \
+    --no-trust-builder-keys \
````
`--no-trust-builder-keys` is no-longer a thing? Arguments here need re-ordering as well.

fanquake commented on contrib/verifybinaries/README.md:
```diff
@@ -1,30 +1,89 @@
+./contrib/verifybinaries/verify.py pub 22.0-x86 \
+    --no-trust-builder-keys \
+    --trusted-keys 74E2DEF5D77260B98BC19438099BAD163C70FBFA,9D3CC86A72F8494342EA5FD10A41BDC3F4FAFF1C \
+    --min-trusted-sigs 10
```
```suggestion
    --min-good-sigs 10
```

fanquake reviewed (commented):
Did some basic testing that this "works". I wouldn't like to hold merging this up past a good-enough state, so happy for comments/nits/follows to be addressed soon after.

theuni committed:
verifybinaries: use recommended keyserver by default

theuni commented:
Thanks for the review @fanquake, comments addressed.

LMK if you'd like me to squash. I could either squash mine into the original authors' or mine all into one "fixup" on top of theirs.

fanquake commented:
@theuni I don't particularly mind. I think the commits here look ok as-is. If you want to squash them all down into a fixup, that's probably also ok, but going to be a bit messy. Squashing all the changes back into the original commits is probably going to be more messy again.

dergoegge commented on contrib/verifybinaries/README.md:
````diff
@@ -1,30 +1,88 @@
+Get JSON output and don't prompt for user input (no auto key import):
+
+```sh
+./contrib/verifybinaries/verify.py pub 22.0-x86 --json
````
```suggestion
./contrib/verifybinaries/verify.py --json pub 22.0-x86
```

otherwise i get this:
```
$ ./contrib/verifybinaries/verify.py pub 22.0-x86 --json
usage: verify.py [-h] [-v] [-q] [--import-keys] [--min-good-sigs [MIN_GOOD_SIGS]] [--keyserver [KEYSERVER]] [--trusted-keys [TRUSTED_KEYS]] [--json] {pub,bin} ...
verify.py: error: unrecognized arguments: --json
```

theuni replied:
Thanks. I'll squashed this into another commit.

josibake reviewed (commented):
Concept ACK

Downloading for a specific platform doesn't work, so I'd suggest we either fix it before merging or remove it from the docs. I also left a few other nits but they aren't important.

josibake commented on contrib/verifybinaries/test.py:
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env python3
```
+1

theuni replied:
Makes sense to me but I'd rather keep the changes down in this PR. Let's do as a follow-up?

TheCharlatan commented on contrib/verifybinaries/verify.py:
```diff
@@ -47,137 +115,598 @@ def parse_version_string(version_str):
+        return sigs_status
+
+    # Extract hashes and filenames
+    hashes_to_verify = parse_sums_file(SUMS_FILENAME, os_filter)
```
```suggestion
    hashes_to_verify = parse_sums_file(SUMS_FILENAME, [os_filter])
```

TheCharlatan replied:
`parse_sums_file` expects a list of strings. This should fix the download selection filter.

josibake replied:
confirmed this does fix the download selection filter

theuni replied:
Thanks!

TheCharlatan commented:
Nit: After discovering https://github.com/bitcoin/bitcoin/pull/27358#discussion_r1159606675 I ran `mypy verify.py` and:
```
verify.py:129: error: Module has no attribute "HTTPError"
verify.py:588: error: Argument 1 to "verify_shasums_signature" has incompatible type "Path"; expected "str"
Found 2 errors in 1 file (checked 1 source file)
```

josibake commented on contrib/verifybinaries/verify.py:
```diff
@@ -47,137 +115,598 @@ def parse_version_string(version_str):
     for _, binary_filename in hashes_to_verify:
-        print(f"Downloading {binary_filename}")
-        download_with_wget(HOST1 + remote_dir + binary_filename)
+        log.info(f"downloading {binary_filename}")
```
nit: it would be nice to have it also print out the destination, e.g "downloading {binary_filename} to {destination}"

fanquake replied:
added to 27440.

theuni committed:
verifybinaries: fix OS download filter

Co-authored-by: Reproducibility Matters <seb.kung@gmail.com>

theuni committed:
verifybinaries: catch the correct exception

theuni committed:
verifybinaries: fix argument type error pointed out by mypy

theuni commented:
Thanks for the review @TheCharlatan and @josibake. I've addressed most of the feedback with the exception of a few nits as I don't want this to drag out forever.

I squashed some of the readme changes to keep the number of commits down. The diff from the previous changes can be seen with: `git diff 96344222f3..754fb6bb81` .

josibake commented:
ACK https://github.com/bitcoin/bitcoin/pull/27358/commits/754fb6bb8125317575edec7c20b5617ad27a9bdd

fanquake commented:
We can have a followup, to address any further issues/improvements.

Sjors commented:
Followup suggestions:

1. allow user to specify the domain, e.g. `--domain=bitcoincore.org` so you don't get warnings about bitcoin.org, which only has up to version 22.0 at the moment. Conversely, it will make a future domain evacuation safer.

2. "osx" in the documentation example should be "macos" (but it doesn't work either way as @theuni found)

3. "The os/arch filters like verify.py 22.0-osx no longer work for me and I'm not quite sure why." - would be nice to restore this, or just file a Github issue and remove it from the docs

theuni commented:
@Sjors That filter problem should've been fixed by https://github.com/bitcoin/bitcoin/pull/27358/commits/4b23b488d2c5662215d78e4963ef5a2b86b4e25b, my comment was from before then. Is what's now in master still broken for you?

Sjors commented:
@theuni yes. On Intel macOS 13.3:

```
 % ./contrib/verifybinaries/verify.py pub 24.0.1-osx 
…
[ERROR] no files matched the platform specified
```

Ditto for `24.0.1-macos`

theuni commented:
@Sjors the filter works almost like a grep, so it depends on the release. Starting with v23 the apple binaries were renamed.

fanquake commented on contrib/verifybinaries/README.md:
````diff
@@ -1,30 +1,88 @@
+    --min-good-sigs 10 pub 22.0-x86
 ```
 
 If you only want to download the binaries of certain platform, add the corresponding suffix, e.g.:
````
See 27440.

fanquake commented on contrib/verifybinaries/test.py:
```diff
@@ -0,0 +1,59 @@
+    expect_code(run_verify("", "pub", '0.32.awefa.12f9h'), 11, "Malformed version should fail")
+    expect_code(run_verify('--min-good-sigs 20', "pub", "22.0"), 9, "--min-good-sigs 20 should fail")
+
+    print("- testing multisig verification (22.0)", flush=True)
```
done in 27440

fanquake commented:
Covered most of the followup requests in #27440.
